### ISSUE TITLE
it seems jack1 cannot be used

### ISSUE THREAD

**Issue by magnetophon** [NixOS/nixpkgs#11079]
@wkennington: I just installed jack1 (and uninstalled jack2), and it doesn't start from qjackctl.
It does start from cli, but other programs don't seem to see it.  

~~I don't understand how other programs COULD see it, as they're all built with (lib)jack2.~~

~~I discussed this with @henrytill on IRC, and we came to the conclusion that it could only work if all occurrences of jack2 are replaced by jack, and in all-pkgs.nix, jack=jack2;
Then in configuration.nix, one could override jack to be jack1~~ 
EDITED: Discussion below shows it shouldn't matter that we compile against jack2.

What are your thoughts on this?

**Comment by magnetophon** [comment]
@wkennington:  You replied on IRC that jack1 is not supported on NixOS.
If that is so, why do we even have it?

You also asked if jack1 has advantages over jack2.
There are a couple of trade-offs:
https://github.com/jackaudio/jackaudio.github.com/wiki/Q_difference_jack1_jack2

Also from that page: "Programs compiled against Jack 1 will work with Jack 2 without recompile (and vice versa)."
So that answers my question on whether it even COULD work as is on NixOS.

**Comment by vcunat** [comment]
> Also from that page: "Programs compiled against Jack 1 will work with Jack 2 without recompile (and vice versa)."

Working without recompiling everything seems a rather important property to me.

**Comment by vcunat** [comment]
> You replied on IRC that jack1 is not supported on NixOS. If that is so, why do we even have it?

I suspect it's only about noone stepping forward (yet) to maintain the support.

**Comment by magnetophon** [comment]
@vcunat Does the solution proposed above look good to you?
If so, I'd be willing to do the work.
Would it be possible to have Hydra build both versions of each program needing jack?

**Comment by vcunat** [comment]
The solution above would recompile all.

I don't know Jack stuff, but if compiling against jack1 would produce jack2-compatible results, why not do that instead?

**Comment by vcunat** [comment]
Clearly it's about the trade-offs, linked above, and there we have

>  For most folks, it doesn't matter. For others, there is typically one or two features that cause them to prefer one or the other.

That sounds as a reasonable default.

**Comment by vcunat** [comment]
I suppose it's vital to find out why (some of) our current binaries refuse to work with jack1, as the link above seems to claim they should work.

**Comment by magnetophon** [comment]
@vcunat afaik the link above claims it cannot work until you compile the programs with the correct jack version.

**Comment by vcunat** [comment]
How so? It says *Programs compiled against Jack 1 will work with Jack 2 without recompile (and vice versa).*

**Comment by magnetophon** [comment]
@vcunat oops, my fault. I read "will NOT work".  Sorry for the confusion.

**Comment by vcunat** [comment]
:-D Good that we've found that out early.

**Comment by nbp** [comment]
https://github.com/NixOS/nixpkgs/issues/11079#issuecomment-157384804
> So that answers my question […]

From what I understand this thread is closed.  Feel free to re-open and clarify the issue otherwise.

**Comment by magnetophon** [comment]
@nbp 
There was a bit of confusion on my side, and that is now cleared up.
The fact remains that jack1 doesn't work on NixOS, so I don't see why you closed the issue.

Unfortunately I don't have time atm to track down why it doesn't work.
Also I wouldn't really know where to start.

**Comment by nbp** [comment]
@magnetophon Ok, my bad, I thought this was only a question on the fact that it might hypotetically not work, and that this question got answered.

So if this is a bug, we would need a way to reproduce it, do you have any idea how to proceed?

**Comment by magnetophon** [comment]
@nbp 
Reproducing is easy:
- install jack1 instead of jack2
- start up jack (via cli, since qjackctl doesn't work with jack1 either)
- start a (any) program that uses jack
- it will say: cannot connect to jack server.

**Comment by vcunat** [comment]
I edited the description to better reflect the status.

**Comment by oxij** [comment]
I actually debugged this quite a while ago. The problem is that `jack1`
and `jack2` use different paths to control sockets and binaries compiled
with `jack2` try to connect to jack using the wrong address. So their
compatibility claim is void in practice.

Anyway, overriding jack2 to jack1 seemed to work at least for
zynaddsubfx. I didn't play my instrument for a while, so I'm not sure it
still works.

**Comment by magnetophon** [comment]
I asked about this on IRC, and one of the authors of jack ( @x42 ) said:
jack1 and jack2 are 100% ABI compatible.   programs compiled and linked with either will work with the other.

**Comment by x42** [comment]
@oxij  there are no paths to contol sockets and binaries compiled into the jack-clients.
Those are all abstracted by libjack and libjackserver (runtime loaded)

This all sounds like jack1 setup is not correct (maybe missing permissions in /dev/shm or something like this). if it were a  ABI compatibility issue the jack client would not even start (missing symbol, missing library).  

"cannot connect to jack" is a most likley setup error in the jack1 package on the distribution that you're using.

**Comment by x42** [comment]
PS. another explanation is that  the jack application is statically linked against libjack (don't ever do that) or that jack clients use different versions of libjack (more likley on NixOS).  There must always only be one libjack.so around on a system and that lib must used by all applications.  You cannot mix different libjack.so.

**Comment by x42** [comment]
issue identified using `strace -f -e trace=open`

Jack is a client/server mechanism.  All clients must use the same shared-library (libjack.so.0).
This library can be provided by different implementations (ie. jack1, jack2) and there must always only be one.  In traditional unix,  libjack.so is in `/usr/lib` and used by all jack clients.

On NixOS however every client uses its "own" from the /nix/storage/... tree rather than the shared one:

```
$ strace -f -e trace=open jack_capture
[...]
open("/nix/store/l32g7d8xr2vnxdvq211r260nwmmmrprw-libjack2-1.9.10/lib/libjack.so.0", O_RDONLY|O_CLOEXEC) = 3
[...]

$ strace -f -e trace=open jack_lsp
open("/nix/store/yyy0snpj4yhskmd48kvz5npdvzhiql3g-jack1-0.124.1/lib64/libjack.so.0", O_RDONLY|O_CLOEXEC) = 3
[...]
```

This cannot work.

**Comment by magnetophon** [comment]
setting LD_LIBRARY_PATH to point to  jack1 makes it work.
tested with qjackctl and slgui, both of which didn't work before.

**Comment by vcunat** [comment]
Thanks, @x42, for explaining that it's required to use the same `libjack.so` in client and server, even if only minor version numbers differ.

The situation seems very alike to libGL and friends which are symlinked from `/run/opengl-driver{,-32}/lib` to the currently chosen implementation (and that path is in `$LD_LIBRARY_PATH`). We could use the same mechanism; we could even put libjack.so next to libGL.so &ndash; it would be a misnomer though. Using a single dir for `$LD_LIBRARY_PATH` still seems to have some advantages, as we have VDPAU in there already, for example, and who knows how many similar "driver" designs will be discovered in future.

**Comment by magnetophon** [comment]
The reason I originally installed jack1 was to see if it would fix an issue with my interface (xruns at low latency and high dsp loads).
It does, so I'd really like to get this fixed.
@vcunat
Do you think the solution you proposed above is the right one? (you sound unsure to me)
If ys, I can try and implement it. (I don't understand it yet, but I haven't looked into it yet either)

**Comment by vcunat** [comment]
I'm pretty certain that we should have a solution along those lines; I'm unsure about some details, such as whether it should be put into `/run/jack` or somewhere else.

In more detail:
- `$LD_LIBRARY_PATH` should get extended by `:/run/jack/lib:run/jack-32/lib` by NixOS, perhaps conditionally, based on whether some jack is enabled. There might not be such need as for libGL to support having both 32-bit and 64-bit version at once, but if it's possible for the jack server, it's better to be prepared to eventually support such a scenario via an option (like `hardware.opengl.driSupport32Bit`), and for that we need to have separate paths to avoid the name clash on `libjack.so`.
- When a jack server starts, it should setup those `/run/jack*` paths and symlink the correct `libjack.so*` into it. This will ensure that newly run applications will get the correct one even if you e.g. had both jack server versions in your config and always stopped one and started the other.

I'll leave implementation and testing to you/others ;-) I've never used jack anyway... but feel free to ask.

**Comment by x42** [comment]
@vcunat ,  good call.  The situation is indeed not unlike  kernel-driver <> user-space-lib (openGL, ALSA).

The "jack-server <> libjack" situation is exactly like "X-server  <> libX11,libxcb".  How is that handled on NixOS?

in case of x11/xorg the protocol (lib to server) is versioned so that different libs can communicate with a different server version, however in reality this is not always given.  (e.g Ubuntu regularly breaks this: 3rd party apps statically linked with libX11 or shipping their own libX11 fail to work).

@magnetophon: The mechanism vcunat describes is the correct one. The "uncertainty" is where to put it. `/run/opengl-driver/` is definitely not the correct location. It'll work as a quick hack (since all the infrastructure and ld-paths for that are already in place).
Even `/run/` may not be ideal, since this is a user-space/user-space communication issue and not a  kernel/use-space one. Anyway that's now down to NixOS's interpretation of the FHS.

**Comment by x42** [comment]
"When a jack server starts, it should setup those /run/jack* paths and symlink the correct"

That is not an option. jack-clients can auto-start the server. The mechanism is provided by libjack.

jack-client loads libjack... libjack tries to connect to server..  if the server is not running -> auto-start (libjack2 uses dbus for that, libjack1 a custom mechanism to launch the server).

**Comment by vcunat** [comment]
> in case of x11/xorg the protocol (lib to server) is versioned so that different libs can communicate with a different server version, however in reality this is not always given.

We were always letting packages use those X libraries that they were built against. So far `libGL` and related libs are the only significant cases where we do such re-linking during runtime (to allow proprietary drivers). I haven't noticed anyone having problems with X due to this, but we have significantly smaller user base than Ubuntu. (I only occasionally run packages linked many months ago while keeping the often-used packages fairly up to date.)

> That is not an option. jack-clients can auto-start the server. The mechanism is provided by libjack.

I'm perhaps too active in things I don't understand (jack) ;-) I assumed it's a daemon that's run globally by init (perhaps socket-actived or such). It will likely need more thought then...

**Comment by x42** [comment]
jack is run as user (not as root). It's a sound-server not unlike pulseaudio which is also started by the user (or the desktop session).

**Comment by magnetophon** [comment]
@vcunat @ x42
I'd like to get this working, but I nead some guidance.
Coud you guys eli5 the above?  ;)

**Comment by x42** [comment]
@magnetophon  the idea was that when jackd is starting it would set up the symlinks required for all jack-clients...  while that's a great idea in general, it's not feasible to do that with jack: 
 - because jackd does not run as root (and hence can't create symlinks system-wide)
 - because jack-clients can start jackd. (the server is not always the first)

**Comment by Profpatsch** [comment]
(triage) any news on this issue?

**Comment by Profpatsch** [comment]
(triage) is it still broken? fixed?

**Comment by oxij** [comment]
AFAIK that's still an issue and I don't see any simple solutions except `LD_LIBRARY_PATH`.

Also, personally, I have a problem with autostarting the daemon in the first place. I think daemons are to be managed by service files, not by idiosyncratic startup login in libraries and dbus.

**Comment by luzpaz** [comment]
This should probably be closed as jack1 is obsolete

**Comment by magnetophon** [comment]
> This should probably be closed as jack1 is obsolete

Thanks for the heads up!
